hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestNameNodeRpcServer.java

```diff
@@ -59,5 +74,67 @@ public void testNamenodeRpcBindAny() throws IOException {
       conf.unset(DFS_NAMENODE_RPC_BIND_HOST_KEY);
     }
   }
+
+  /**
+   * A test to make sure that if an authorized user adds "clientIp:" to their
+   * caller context, it will be used to make locality decisions on the NN.
+   */
+  @Test
+  public void testNamenodeRpcClientIpProxy()
+      throws InterruptedException, IOException {
+    Configuration conf = new HdfsConfiguration();
+
+    conf.set(DFS_NAMENODE_IP_PROXY_USERS, "fake_joe");
```

Comment:
That isn't an easy test to write, because in the mini dfs cluster, all of the datanodes are on the same ip. So without the clientIp being forced, the namenode will pick a datanode randomly. That is why I made three datanodes, so the test will only accidentally pass 1/3 of the time.

Reply:
I could add multiple trials. Doing the math 1/3 ^ 10 = 1.7e-5 and 1/3 ^ 20 = 2.9e-10, so 20 trials will bring the failure rate down to 2.9e-10. That seems doable.

Reply:
makes sense to me.